File: sources/entity/tests/test_confidence_downgrade.php
<?php
/**
 * Test suite for confidence auto-downgrade after failed re-validation.
 * Run: php tests/test_confidence_downgrade.php
 *
 * Tests the logic that downgrades confidence to 'low' and adds a
 * validation_warning when Phase 8 re-validation also fails.
 */

$passed = 0;
$failed = 0;

function test(string $label, $expected, $actual): void
{
    global $passed, $failed;
    if ($expected === $actual) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label}\n    Expected: " . json_encode($expected) . "\n    Got:      " . json_encode($actual) . "\n";
        $failed++;
    }
}

/**
 * Simulate the confidence downgrade logic from lookup.php (lines 148-154).
 * $needsReanalysis: whether Phase 8 was triggered
 * $rvStatus: registry_validation status after Phase 8
 * $originalConfidence: the confidence from the LLM report
 */
function applyDowngrade(bool $needsReanalysis, ?string $rvStatus, string $originalConfidence): array
{
    $report = [
        'confidence' => $originalConfidence,
        'registry_validation' => $rvStatus ? ['status' => $rvStatus] : null,
    ];

    // Mirror the exact logic from lookup.php
    $rvStatusCheck = $report['registry_validation']['status'] ?? null;
    if ($needsReanalysis && $rvStatusCheck && $rvStatusCheck !== 'verified') {
        $report['confidence'] = 'low';
        $report['validation_warning'] = 'Registry validation failed after re-analysis — confidence auto-downgraded';
    }

    return $report;
}

echo "=== Confidence Auto-Downgrade Tests ===\n\n";

// 1. Re-analysis triggered, re-validation failed with name_mismatch → downgrade
echo "1. Re-analysis + name_mismatch → downgrade to low:\n";
$r = applyDowngrade(true, 'name_mismatch', 'high');
test('Confidence downgraded to low', 'low', $r['confidence']);
test('Has validation_warning', true, isset($r['validation_warning']));
echo "\n";

// 2. Re-analysis triggered, re-validation failed with name_match_bad_status → downgrade
echo "2. Re-analysis + name_match_bad_status → downgrade to low:\n";
$r = applyDowngrade(true, 'name_match_bad_status', 'medium');
test('Confidence downgraded to low', 'low', $r['confidence']);
test('Has validation_warning', true, isset($r['validation_warning']));
echo "\n";

// 3. Re-analysis triggered, re-validation verified → NO downgrade
echo "3. Re-analysis + verified → no downgrade:\n";
$r = applyDowngrade(true, 'verified', 'high');
test('Confidence stays high', 'high', $r['confidence']);
test('No validation_warning', false, isset($r['validation_warning']));
echo "\n";

// 4. No re-analysis needed, validation failed → NO downgrade (Phase 7 only)
echo "4. No re-analysis + failed validation → no downgrade:\n";
$r = applyDowngrade(false, 'name_mismatch', 'medium');
test('Confidence stays medium', 'medium', $r['confidence']);
test('No validation_warning', false, isset($r['validation_warning']));
echo "\n";

// 5. Re-analysis triggered, no registry validation at all → NO downgrade
echo "5. Re-analysis + no registry validation → no downgrade:\n";
$r = applyDowngrade(true, null, 'high');
test('Confidence stays high', 'high', $r['confidence']);
test('No validation_warning', false, isset($r['validation_warning']));
echo "\n";

// 6. Already low confidence + re-analysis failed → stays low with warning
echo "6. Already low + re-analysis failed → stays low with warning:\n";
$r = applyDowngrade(true, 'name_mismatch', 'low');
test('Confidence is low', 'low', $r['confidence']);
test('Has validation_warning', true, isset($r['validation_warning']));
echo "\n";

// 7. Re-analysis + fictitious_name status → downgrade
echo "7. Re-analysis + fictitious_name → downgrade:\n";
$r = applyDowngrade(true, 'fictitious_name', 'high');
test('Confidence downgraded to low', 'low', $r['confidence']);
test('Has validation_warning', true, isset($r['validation_warning']));
echo "\n";

// 8. Re-analysis + branch_registration → downgrade
echo "8. Re-analysis + branch_registration → downgrade:\n";
$r = applyDowngrade(true, 'branch_registration', 'medium');
test('Confidence downgraded to low', 'low', $r['confidence']);
test('Has validation_warning', true, isset($r['validation_warning']));
echo "\n";

echo "=== RESULTS ===\n";
echo "Passed: {$passed}\n";
echo "Failed: {$failed}\n";
exit($failed > 0 ? 1 : 0);

File: sources/entity/app/tests/test_confidence_downgrade.py
"""
Faithful Python port of php/tests/test_confidence_downgrade.php.

Tests the confidence auto-downgrade logic that runs after a failed Phase 8
re-validation. In the Python port this lives inline in EntityLookup.run()
(agent.py lines ~181-185):

    rv_status = (report.get('registry_validation') or {}).get('status')
    if needs_reanalysis and rv_status and rv_status != 'verified':
        report['confidence'] = 'low'
        report['validation_warning'] = 'Registry validation failed after re-analysis — confidence auto-downgraded'

Like the PHP test (which re-implements the block as applyDowngrade), this test
replicates that exact block as a local helper and drives the same 8 cases. Pure/offline.
"""

WARNING_MSG = "Registry validation failed after re-analysis — confidence auto-downgraded"


def apply_downgrade(needs_reanalysis, rv_status, original_confidence):
    """Mirror of the inline downgrade block in EntityLookup.run()."""
    report = {
        "confidence": original_confidence,
        "registry_validation": {"status": rv_status} if rv_status else None,
    }

    rv_status_check = (report.get("registry_validation") or {}).get("status")
    if needs_reanalysis and rv_status_check and rv_status_check != "verified":
        report["confidence"] = "low"
        report["validation_warning"] = WARNING_MSG

    return report


# 1. Re-analysis + name_mismatch → downgrade to low
def test_reanalysis_name_mismatch_downgrades():
    r = apply_downgrade(True, "name_mismatch", "high")
    assert r["confidence"] == "low"
    assert "validation_warning" in r


# 2. Re-analysis + name_match_bad_status → downgrade to low
def test_reanalysis_name_match_bad_status_downgrades():
    r = apply_downgrade(True, "name_match_bad_status", "medium")
    assert r["confidence"] == "low"
    assert "validation_warning" in r


# 3. Re-analysis + verified → NO downgrade
def test_reanalysis_verified_no_downgrade():
    r = apply_downgrade(True, "verified", "high")
    assert r["confidence"] == "high"
    assert "validation_warning" not in r


# 4. No re-analysis + failed validation → NO downgrade
def test_no_reanalysis_no_downgrade():
    r = apply_downgrade(False, "name_mismatch", "medium")
    assert r["confidence"] == "medium"
    assert "validation_warning" not in r


# 5. Re-analysis + no registry validation → NO downgrade
def test_reanalysis_no_registry_validation_no_downgrade():
    r = apply_downgrade(True, None, "high")
    assert r["confidence"] == "high"
    assert "validation_warning" not in r


# 6. Already low + re-analysis failed → stays low with warning
def test_already_low_reanalysis_failed_stays_low_with_warning():
    r = apply_downgrade(True, "name_mismatch", "low")
    assert r["confidence"] == "low"
    assert "validation_warning" in r


# 7. Re-analysis + fictitious_name → downgrade
def test_reanalysis_fictitious_name_downgrades():
    r = apply_downgrade(True, "fictitious_name", "high")
    assert r["confidence"] == "low"
    assert "validation_warning" in r


# 8. Re-analysis + branch_registration → downgrade
def test_reanalysis_branch_registration_downgrades():
    r = apply_downgrade(True, "branch_registration", "medium")
    assert r["confidence"] == "low"
    assert "validation_warning" in r
